src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+      }
+    }
+
+    _buffer.pop_front();
+  }
+  assert(_buffer.size() < _buffer_max_size, "_buffer is over-sized.");
+  _buffer.push_back(msg);
+
+  // notify asynclog thread if occupancy is over 3/4
+  size_t sz = _buffer.size();
+  if (sz > (_buffer_max_size >> 2) * 3 ) {
+    _lock.notify();
+  }
+}
+
+void LogAsyncFlusher::enqueue(LogFileOutput& output, const LogDecorations& decorations, const char* msg) {
+  AsyncLogMessage m(output, decorations, os::strdup(msg));
+
+  { // critical area
+    // The rank of _lock is same as _tty_lock on purpuse.
+    // if logging thread is holding _tty_lock now, temporarily yield to _lock.
+    ttyUnlocker ttyul;
+    MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+    enqueue_impl(m);
+  }
+}
+
+// LogMessageBuffer consists of a multiple-part/multiple-line messsages.
+// the mutex here gurantees its integrity.
+void LogAsyncFlusher::enqueue(LogFileOutput& output, LogMessageBuffer::Iterator msg_iterator) {
+  ttyUnlocker ttyul;
+  MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+
+  for (; !msg_iterator.is_at_end(); msg_iterator++) {
+    AsyncLogMessage m(output, msg_iterator.decorations(), os::strdup(msg_iterator.message()));
+    enqueue_impl(m);
+  }
+}
+
+LogAsyncFlusher::LogAsyncFlusher()
+  : _state(ThreadState::Running),
+    _lock(Mutex::tty, "async-log-monitor", true /* allow_vm_block */, Mutex::_safepoint_check_never),
+    _stats(17 /*table_size*/) {
+  if (os::create_thread(this, os::asynclog_thread)) {
+    os::start_thread(this);
+  }
+
+  log_info(logging)("The maximum entries of AsyncLogBuffer: " SIZE_FORMAT ", estimated memory use: " SIZE_FORMAT " bytes",
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+bool AsyncLogMapIterator::do_entry(LogFileOutput* output, uint32_t* counter) {
+  using dummy = LogTagSetMapping<LogTag::__NO_TAG>;
+  LogDecorations decorations(LogLevel::Warning, dummy::tagset(), output->decorators());
+  const int sz = 128;
+  char out_of_band[sz];
+
+  if (*counter > 0) {
+    jio_snprintf(out_of_band, sz, UINT32_FORMAT_W(6) " messages dropped due to async logging", *counter);
+    output->write_blocking(decorations, out_of_band);
+    *counter = 0;
+  }
+
+  return true;
+}
+
+void LogAsyncFlusher::writeback(const LinkedList<AsyncLogMessage>& logs) {
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  while (!it.is_empty()) {
+    AsyncLogMessage* e = it.next();
+    e->writeback();
+  }
+}
+
+void LogAsyncFlusher::flush(bool with_lock) {
+  LinkedListImpl<AsyncLogMessage, ResourceObj::C_HEAP, mtLogging> logs;
+
+  if (with_lock) { // critical area
+    // Caveat: current thread must not hold _tty_lock or other lower rank lockers.
+    // Cannot install ttyUnlocker here because flush() may be invoked before defaultStream
+    // initialization.
+    MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+    _buffer.pop_all(&logs);
+    AsyncLogMapIterator iter;
+    _stats.iterate(&iter);
```

Comment:
All this is really difficult to understand for the casual reader. Clearer naming would help:
- something like "move-list-content" instead of pop_all
- the whole point of this iteration thing is to print the dropped counters, right? A function "Print Drop Counters" would be clearer. Unless you go the one-async-buffer-per-logfileoutput-route I mentioned earlier.

Reply:
I will send out a patch to fix this along with the new discard policy. 
There's another issue.  `_stats.iterate(&iter)` here invokes `output->write_blocking` while holding lock!
This has compromised non-blocking protocol.